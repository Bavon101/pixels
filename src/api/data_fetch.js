import {
  fetchDataBegin, fetchDataFailure,
  fetchDataSuccess, imageFetchSuccess,
} from './data_fetch_statuses';

// Handle HTTP errors since fetch won't.
function handleErrors(response) {
  if (!response.ok) {
    throw Error(response.statusText);
  }
  return response;
}

export default function fetchImages() {
  return (dispatch) => {
    dispatch(fetchDataBegin());
    return fetch('https://pixabay.com/api/?key=12918753-88ae9b08d6ddff4b548ea3571&q=art+blue&image_type=photo&pretty=true&per_page=100', {
      method: 'GET',
    })
      .then(handleErrors)
      .then((res) => res.json())
      .then((images) => {
        dispatch(fetchDataSuccess(images));
        return images;
      })
      .catch((error) => dispatch(fetchDataFailure(error)));
  };
}

export function fetchSingleImage(id) {
  return (dispatch) => {
    dispatch(fetchDataBegin());
    return fetch(`https://pixabay.com/api/?key=12918753-88ae9b08d6ddff4b548ea3571&id=${id}&image_type=photo&pretty=true`, {
      method: 'GET',
    })
      .then(handleErrors)
      .then((res) => res.json())
      .then((image) => {
        dispatch(imageFetchSuccess(image));
        return image;
      })
      .catch((error) => dispatch(fetchDataFailure(error)));
  };
}
