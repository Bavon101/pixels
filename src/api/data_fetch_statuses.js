export const FETCH_DATA_BEGIN = 'api/images/FETCH_DATA_BEGIN';
export const FETCH_DATA_SUCCESS = 'api/images/FETCH_DATA_SUCCESS';
export const FETCH_DATA_FAILURE = 'api/images/FETCH_DATA_FAILURE';
export const IMAGE_FETCH_BEGIN = 'api/images/IMAGE_FETCH_BEGIN';
export const IMAGE_FETCH_SUCCESS = 'api/images/IMAGE_FETCH_SUCCESS';
export const IMAGE_FETCH_FAILURE = 'api/images/IMAGE_FETCH_FAILURE';

export const fetchDataBegin = () => ({
  type: FETCH_DATA_BEGIN,
});

export const fetchDataSuccess = (data) => ({
  type: FETCH_DATA_SUCCESS,
  payload: data,
});

export const fetchDataFailure = (error) => ({
  type: FETCH_DATA_FAILURE,
  payload: error,
});

export const fetchImageBegin = () => ({ type: IMAGE_FETCH_BEGIN });
export const imagefetchFailure = (error) => ({ type: IMAGE_FETCH_FAILURE, payload: error });
export const imageFetchSuccess = (data) => ({ type: IMAGE_FETCH_SUCCESS, payload: data });
