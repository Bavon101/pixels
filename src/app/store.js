import { applyMiddleware, combineReducers, createStore } from '@reduxjs/toolkit';
import thunk from 'redux-thunk';
import imagesReducer from '../redux/images';

const rootReducer = combineReducers(
  {
    imagesReducer,
  },
);
const store = createStore(rootReducer, applyMiddleware(thunk));

export default store;
