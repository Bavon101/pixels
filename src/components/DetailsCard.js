import React from 'react';
import PropTypes from 'prop-types';
import './details_card.css';

export default function DetailsCard(props) {
  const { title, value, light } = props;
  return (
    <div className="card" style={{ backgroundColor: light ? '#4167AF' : '#3F62A6' }}>
      <h3>
        {title}
      </h3>
      <h4>
        {value}
      </h4>
    </div>
  );
}
DetailsCard.propTypes = {
  title: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  light: PropTypes.bool.isRequired,
};
