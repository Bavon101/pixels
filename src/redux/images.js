import {
  FETCH_DATA_BEGIN, FETCH_DATA_FAILURE, FETCH_DATA_SUCCESS,
  IMAGE_FETCH_BEGIN, IMAGE_FETCH_FAILURE, IMAGE_FETCH_SUCCESS,
} from '../api/data_fetch_statuses';

export function imagesReducer(state = {
  images: [],
  loading: false,
  error: null,
}, action) {
  switch (action.type) {
    case FETCH_DATA_BEGIN:

      return {
        ...state,
        loading: true,
      };
    case FETCH_DATA_SUCCESS:
      return {
        ...state,
        loading: false,
        images: [
          ...action.payload.hits.sort((a, b) => b.likes - a.likes),
        ],
      };
    case FETCH_DATA_FAILURE:
      return {
        ...state,
        loading: false,
        error: action.payload,
      };
    default:
      return state;
  }
}

export function imageReducer(state = {
  loading: false,
  image: null,
  error: null,
}, action) {
  switch (action.type) {
    case IMAGE_FETCH_BEGIN:

      return {
        ...state,
        loading: true,
      };
    case IMAGE_FETCH_FAILURE:
      return {
        ...state,
        error: action.payload,
      };
    case IMAGE_FETCH_SUCCESS:
      return {
        ...state,
        image: action.payload.hits[0],
      };
    default:
      return state;
  }
}
