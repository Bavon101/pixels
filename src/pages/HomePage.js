import {
  ImageList, ImageListItem,
  ImageListItemBar, Card, CardMedia, ButtonGroup, Button,
} from '@mui/material';
import React, { useEffect } from 'react';
import ArrowCircleDownRoundedIcon from '@mui/icons-material/ArrowCircleDownRounded';
import FavoriteIcon from '@mui/icons-material/Favorite';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import fetchImages from '../api/data_fetch';
import '../components/image/image_card.css';
import NavgationBar from '../components/NavgationBar';
import Loading from '../components/Loading';
import { imagesFiltered, resetFilters } from '../api/data_fetch_statuses';

export default function HomePage() {
  const imagesData = useSelector((state) => state.imagesReducer);
  const dispatch = useDispatch();
  const images = imagesData.filter;
  useEffect(() => {
    if (imagesData.images.length === 0) {
      dispatch(fetchImages());
    }
  }, []);
  const filter = (filter) => dispatch(imagesFiltered(filter));
  if (images.length === 0) {
    return (
      <div>
        <Loading />
      </div>
    );
  }
  const btnStyle = {
    backgroundColor: '#3F62A6',
  };
  return (
    <div className="home">
      <NavgationBar title="PIXELS" goBack={false} />
      <div style={{ marginTop: '63px', borderRadius: '0' }}>
        <div className="filiter-card">
          <i className="material-icons">filter</i>
          <h5>
            Filter Images with:
          </h5>
        </div>
        <ButtonGroup variant="contained" aria-label="outlined primary button group" sx={{ marginBottom: '10px' }}>
          <Button onClick={() => filter('likes')} sx={imagesData.filter_tag !== 'likes' ? btnStyle : {}}>
            Likes
          </Button>
          <Button onClick={() => filter('views')} sx={btnStyle}>views</Button>
          <Button onClick={() => filter('comments')} sx={btnStyle}>Comments</Button>
          <Button onClick={() => dispatch(resetFilters())} sx={btnStyle}>Clear</Button>
        </ButtonGroup>
        <div className="image-tags">
          <h4>
            {
              images[0].tags.toUpperCase()
            }
          </h4>
          <h5>
            {
              `${images[0].likes.toLocaleString()} ${imagesData.filter_tag}`
            }
          </h5>
        </div>
        <Card component={Link} to={`./details?id=${images[0].id}`}>
          <CardMedia
            component="img"
            height="194"
            image={images[0].webformatURL}
            alt={images[0].tags}
          />
        </Card>
      </div>
      <h4 className="filter-label">
        {`PIXELS FILTERED BY ${imagesData.filter_tag.toUpperCase()}`}
      </h4>
      <ImageList cols={2} gap={8} sx={{ marginTop: '5px' }}>
        {images.map((image) => (
          <ImageListItem component={Link} to={`./details?id=${image.id}`} key={image.id}>
            <ArrowCircleDownRoundedIcon sx={{ color: 'white' }} className="arrow-btn" style={{ transform: 'rotate(-90deg)' }} />
            <div className="image-tags-con">
              <h5>
                {
                  image.tags.toUpperCase().split(',')[0]
                }
              </h5>
            </div>
            <img
              src={`${image.webformatURL}?w=164&h=164&fit=crop&auto=format`}
              srcSet={`${image.webformatURL}?w=164&h=164&fit=crop&auto=format&dpr=2 2x`}
              alt={image.tags}
              loading="lazy"
            />
            <ImageListItemBar
              sx={{
                background:
                  'linear-gradient(to bottom, rgba(0,0,0,0.7) 0%, '
                  + 'rgba(0,0,0,0.3) 70%, rgba(0,0,0,0) 100%)',
              }}
              title=""
              position="bottom"
              actionIcon={(
                <div className="info-conatiner">
                  <FavoriteIcon sx={{ color: 'white' }} />
                  <h6>
                    {image[imagesData.filter_tag].toLocaleString()}
                  </h6>
                </div>
              )}
              actionPosition="right"
            />
          </ImageListItem>
        ))}
      </ImageList>
    </div>
  );
}
