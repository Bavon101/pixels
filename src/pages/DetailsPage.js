import { Card, CardMedia } from '@mui/material';
import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useLocation } from 'react-router-dom';
import { fetchSingleImage } from '../api/data_fetch';
import DetailsCard from '../components/DetailsCard';
import Loading from '../components/Loading';
import NavgationBar from '../components/NavgationBar';
import './styles.css';

export default function DetailsPage() {
  const { search } = useLocation();
  const id = new URLSearchParams(search).get('id');
  const image = useSelector((state) => state.imageReducer);
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(fetchSingleImage(id));
  }, []);
  if (image.image == null) {
    return (
      <div>
        <Loading />
      </div>
    );
  }
  return (
    <div>
      <NavgationBar title="Pixel Data " goBack />
      <div style={{ marginTop: '63px', borderRadius: '0' }}>
        <div className="image-tags">
          <h4>
            {
              image.image.tags.toUpperCase()
          }
          </h4>
          <h5>
            {
              `${image.image.likes.toLocaleString()} likes`
            }
          </h5>
        </div>
        <Card>
          <CardMedia
            component="img"
            height="194"
            image={image.image.webformatURL}
            alt={image.image.tags}
          />
        </Card>
      </div>
      <DetailsCard title="Views" value={`${image.image.views.toLocaleString()} `} icon="visibility" />
      <DetailsCard title="Downloads" value={`${image.image.downloads.toLocaleString()} `} light icon="file_download" />
      <DetailsCard title="Collections" value={`${image.image.collections.toLocaleString()} `} icon="collections" />
      <DetailsCard title="Comments" value={`${image.image.comments.toLocaleString()} `} light icon="chat_bubble" />
      <DetailsCard title="Likes" value={`${image.image.likes.toLocaleString()} `} icon="mood" />
    </div>
  );
}
