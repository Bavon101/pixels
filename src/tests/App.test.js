import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter, Route, Routes } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { fetchDataSuccess, imageFetchSuccess } from '../api/data_fetch_statuses';
import store from '../app/store';
import DetailsCard from '../components/DetailsCard';
import DetailsPage from '../pages/DetailsPage';
import HomePage from '../pages/HomePage';
import { imageReducer, imagesReducer } from '../redux/images';
import mockImages from './mock_data/images';
import mockImage from './mock_data/image';

it('Render HomePage', () => {
  const page = renderer.create(
    <Provider store={store}>
      <HomePage />
    </Provider>,
  ).toJSON();
  expect(page).toMatchSnapshot();
});

it('Render Details Page', () => {
  const page = renderer.create(
    <Provider store={store}>
      <BrowserRouter>
        <Routes>
          <Route path="/details?id=63636733" element={<DetailsPage />} />
        </Routes>
      </BrowserRouter>
    </Provider>,
  ).toJSON();
  expect(page).toMatchInlineSnapshot('null');
});

it('Render Details card', () => {
  const card = renderer.create(
    <DetailsCard title="Views" value="20" light icon="file_download" />,
  );
  expect(card).toMatchSnapshot();
});

it('Update fetched images', () => {
  const state = imagesReducer(undefined, fetchDataSuccess(mockImages));
  expect(state.images.length).toEqual(20);
});

it('Update store with single image', () => {
  expect(imageReducer(undefined, imageFetchSuccess(mockImage)).image).toEqual(mockImage.hits[0]);
});
