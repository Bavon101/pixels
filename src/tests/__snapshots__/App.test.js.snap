// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Details card 1`] = `
<div
  className="card"
  style={
    Object {
      "backgroundColor": "#4167AF",
    }
  }
>
  <h3>
    Views
  </h3>
  <div
    className="value-icon"
  >
    <h4>
      20
    </h4>
    <i
      className="material-icons"
    >
      file_download
    </i>
  </div>
</div>
`;

exports[`Render HomePage 1`] = `
<div>
  <div
    className="loader"
  />
</div>
`;
